lib/worker.js

```diff
@@ -0,0 +1,210 @@
+};
+
+Worker.prototype[installEventsSymbol] = function() {
+  const workerObject = this;
+  const workerContext = this[workerContextSymbol];
+
+  const onerror = function(payload) {
+    var ErrorConstructor = builtinErrorTypes.get(payload.builtinType);
+    if (typeof ErrorConstructor !== 'function')
+      ErrorConstructor = Error;
+    const error = new ErrorConstructor(payload.message);
+    error.stack = payload.stack;
+    util._extend(error, payload.additionalProperties);
+    workerObject.emit('error', error);
+  };
+
+  workerContext._onexit = function(exitCode) {
+    workerObject[workerContextSymbol] = null;
+    workerObject.emit('exit', exitCode);
+  };
+
+  workerContext._onmessage = function(payload, messageType) {
+    payload = JSONParse(payload);
+    switch (messageType) {
+      case WorkerContextBinding.USER:
+        return workerObject.emit('message', payload);
+      case WorkerContextBinding.INTERNAL:
+        assert.fail('unreachable');
+      case WorkerContextBinding.EXCEPTION:
+        return onerror(payload);
+      default:
+        assert.fail('unreachable');
+    }
+  };
+};
+
+Worker.prototype[checkAliveSymbol] = function() {
+  if (!this[workerContextSymbol])
+    throw new RangeError('this worker has been terminated');
+};
+
+Worker.prototype.postMessage = function(payload) {
+  this[checkAliveSymbol]();
+  this[workerContextSymbol].postMessage(JSONStringify(payload),
+                                        EMPTY_ARRAY,
+                                        WorkerContextBinding.USER);
+};
+
+Worker.prototype.terminate = function(callback) {
+  this[checkAliveSymbol]();
+  var context = this[workerContextSymbol];
+  this[workerContextSymbol] = null;
+  if (typeof callback === 'function') {
+    this.once('exit', function(exitCode) {
+      callback(null, exitCode);
+    });
+  }
+  context.terminate();
+};
+
+Worker.prototype.ref = function() {
+  this[checkAliveSymbol]();
+  this[workerContextSymbol].ref();
+};
+
+Worker.prototype.unref = function() {
+  this[checkAliveSymbol]();
+  this[workerContextSymbol].unref();
+};
+
+if (process.isWorkerInstance) {
+  const postMessage = function(payload, transferList, type) {
+    if (!Array.isArray(transferList))
+      throw new Error('transferList must be an array');
+
+    WorkerContextBinding.workerWrapper._postMessage(JSONStringify(payload),
+                                                    transferList,
+                                                    type);
+  };
+  const workerFatalError = function(er) {
+    const defaultStack = null;
+    const defaultMessage = '[toString() conversion failed]';
+    const defaultBuiltinType = 'Error';
+
+    var message = defaultMessage;
+    var builtinType = defaultBuiltinType;
+    var stack = defaultStack;
+    var additionalProperties = {};
+
+    if (er instanceof Error) {
+      try {
+        builtinType = er.name;
+      } catch (ignore) {}
+
+      if (typeof builtinType !== 'string')
```

Comment:
@thefourtheye That's taken care of by assigning `builtinType = er.name`. At that point `builtinType` can't be a getter.

@bmeck I would always agree with you on that point, except I am on the fence about making sure to prevent errors from being thrown while another error is being generated. On the fence about it being used here. Thoughts on potential pitfalls of this approach?